File: graffiti/src/main/java/com/ue/graffiti/touch/KeepDrawingTouch.java
package com.ue.graffiti.touch;

import android.graphics.PointF;
import android.graphics.drawable.BitmapDrawable;
import android.widget.Toast;

import com.ue.graffiti.R;
import com.ue.graffiti.helper.DialogHelper;
import com.ue.graffiti.widget.CanvasView;

public class KeepDrawingTouch extends Touch {
    private PointF downPoint;
    private KeepDrawingTouchListener mKeepDrawingTouchListener;

    public void setKeepDrawingListener(KeepDrawingTouchListener keepDrawingTouchListener) {
        mKeepDrawingTouchListener = keepDrawingTouchListener;
    }

    public KeepDrawingTouch(CanvasView canvasView) {
        super(canvasView);
        downPoint = new PointF();
    }

    // 第一只手指按下
    @Override
    public void down1() {
        downPoint.set(curPoint);
        updateSavedBitmap(false);
        //调取起始点标志图片
        BitmapDrawable startFlag = (BitmapDrawable) getContext().getResources().getDrawable(R.drawable.img_startflag);
        savedCanvas.drawBitmap(startFlag.getBitmap(), downPoint.x, downPoint.y, null);

        String tip = "您确定以(" + (int) downPoint.x + "," + (int) downPoint.y + ")为起始点并开始重力绘图？";
        DialogHelper.showSensorDrawingDialog(getContext(), tip, (dialog, which) -> {
            if (mKeepDrawingTouchListener != null) {
                mKeepDrawingTouchListener.onDownPoint(downPoint);
                mKeepDrawingTouchListener.registerKeepDrawingSensor();
            }
            Toast.makeText(getContext(), "摆动手机画图吧", Toast.LENGTH_SHORT).show();
        });
    }

    public interface KeepDrawingTouchListener {
        void onDownPoint(PointF downPoint);

        void registerKeepDrawingSensor();
    }
}

File: graffiti/src/main/java/com/ue/graffiti/touch/KeepDrawingTouch.kt
package com.ue.graffiti.touch

import android.content.DialogInterface
import android.graphics.PointF
import android.graphics.drawable.BitmapDrawable
import com.ue.graffiti.R
import com.ue.graffiti.helper.DialogHelper
import com.ue.library.util.toast
import com.ue.graffiti.widget.CanvasView

class KeepDrawingTouch(canvasView: CanvasView) : Touch(canvasView) {
    private val downPoint: PointF = PointF()
    var keepDrawingTouchListener: KeepDrawingTouchListener? = null

    // 第一只手指按下
    override fun down1() {
        downPoint.set(curPoint)
        updateSavedBitmap(false)
        //调取起始点标志图片
        val startFlag = context.resources.getDrawable(R.drawable.img_startflag) as BitmapDrawable
        savedCanvas.drawBitmap(startFlag.bitmap, downPoint.x, downPoint.y, null)

        val tip = context.getString(R.string.gr_begin_sensor_draw, downPoint.x.toInt(), downPoint.y.toInt())
        DialogHelper.showSensorDrawingDialog(context, tip, DialogInterface.OnClickListener { _, _ ->
            keepDrawingTouchListener?.apply {
                onDownPoint(downPoint)
                registerKeepDrawingSensor()
            }
            context.toast(R.string.gr_shake_to_draw)
        })
    }

    interface KeepDrawingTouchListener {
        fun onDownPoint(downPoint: PointF)

        fun registerKeepDrawingSensor()
    }
}